tsdb/executor.go

```diff
@@ -592,24 +592,88 @@ func (e *SelectExecutor) close() {
 	}
 }
 
-func (e *SelectExecutor) processTopBottom(results [][]interface{}, columnNames []string) ([][]interface{}, error) {
-	aggregates := e.stmt.FunctionCalls()
-	var call *influxql.Call
-	process := false
-	for _, c := range aggregates {
-		if c.Name == "top" || c.Name == "bottom" {
-			process = true
-			call = c
-			break
+func (e *SelectExecutor) processFunctions(results [][]interface{}, columnNames []string) ([][]interface{}, error) {
+	callInPosition := e.stmt.FunctionCallsByPosition()
+	hasTimeField := e.stmt.HasTimeFieldSpecified()
+
+	var err error
+	for i, calls := range callInPosition {
+		if len(calls) == 1 {
+			var c *influxql.Call
+			c = calls[0]
+
+			switch c.Name {
+			case "top", "bottom":
+				results, err = e.processAggregates(results, columnNames, c)
+				if err != nil {
+					return results, err
+				}
+			case "first", "last", "min", "max":
+				results, err = e.processSelectors(results, i, hasTimeField, columnNames)
+				if err != nil {
+					return results, err
+				}
+			}
+		}
+	}
+
+	return results, nil
+}
+
+func (e *SelectExecutor) processSelectors(results [][]interface{}, callPosition int, hasTimeField bool, columnNames []string) ([][]interface{}, error) {
+	for i, vals := range results {
+		for j := 1; j < len(vals); j++ {
+			switch v := vals[j].(type) {
+			case PositionPoint:
+				tMin := vals[0].(time.Time)
+				results[i] = e.selectorPointToQueryResult(vals, hasTimeField, callPosition, v, tMin, columnNames)
+			}
+		}
+	}
+	return results, nil
+}
+
+func (e *SelectExecutor) selectorPointToQueryResult(row []interface{}, hasTimeField bool, columnIndex int, p PositionPoint, tMin time.Time, columnNames []string) []interface{} {
+	// if the row doesn't have enough columns, expand it
```

Comment:
It's because the cursor gets back `[][]interface{}`, so for a query like:

```sql
select tx, max(tx) from network
```

It gets back the first row as `[[tx,tx]]` and what it needs to be is `[tx, tx]`.  Basically you have all the data, but the way the map/reduce funcs return data, we have to send back slices of values as a single `column`.  That array needs to be expanded to allow flattening of that slice.